packages/worker/src/workflow/vm-shared.ts

```diff
@@ -0,0 +1,364 @@
+      this.overridePrepareStackTrace();
+      this.setPromiseHook();
+      this.installed = true;
+    }
+  }
+
+  /**
+   * Unset all installed global hooks
+   * TODO(bergundy): Where should this be called?
+   */
+  uninstall(): void {
+    this.stopPromiseHook();
+    Error.prepareStackTrace = this.origPrepareStackTrace;
+    this.installed = false;
+  }
+
+  private overridePrepareStackTrace(): void {
+    const OuterError = Error;
+    // Augment the vm-global Error stack trace prepare function
+    // NOTE: this means that multiple instances of this class in the same VM
+    // will override each other.
+    // This should be a non-issue in most cases since we typically construct a single instance of
+    // this class per Worker thread.
+    // See: https://v8.dev/docs/stack-trace-api#customizing-stack-traces
+    Error.prepareStackTrace = (err, stackTraces) => {
+      const inWorkflowContext = OuterError !== err.constructor;
+      if (this.origPrepareStackTrace && !inWorkflowContext) {
+        return this.origPrepareStackTrace(err, stackTraces);
+      }
+      // Set the currentStackTrace so it can be used in the promise `init` hook below
+      this.currentStackTrace = [];
+      const converted = stackTraces.map((callsite) => {
+        const line = callsite.getLineNumber();
+        const column = callsite.getColumnNumber();
+        const filename = callsite.getFileName();
+        const sourceMapConsumer = filename && this.bundleFilenameToSourceMapConsumer.get(filename);
+        if (sourceMapConsumer && line && column) {
+          const pos = sourceMapConsumer.originalPositionFor({ line, column });
+
+          const name = pos.name || formatCallsiteName(callsite);
+          this.currentStackTrace?.push({
+            filePath: pos.source ?? undefined,
+            functionName: name ?? undefined,
+            line: pos.line ?? undefined,
+            column: pos.column ?? undefined,
+          });
+
+          return name
+            ? `    at ${name} (${pos.source}:${pos.line}:${pos.column})`
+            : `    at ${pos.source}:${pos.line}:${pos.column}`;
+        } else {
+          const name = formatCallsiteName(callsite);
+
+          this.currentStackTrace?.push({
+            filePath: filename ?? undefined,
+            functionName: name ?? undefined,
+            line: line ?? undefined,
+            column: column ?? undefined,
+          });
+          return `    at ${callsite}`;
+        }
+      });
+      return `${err}\n${converted.join('\n')}`;
+    };
+  }
+
+  private setPromiseHook(): void {
+    // Track Promise aggregators like `race` and `all` to link their internally created promises
+    let currentAggregation: Promise<unknown> | undefined = undefined;
+
+    // This also is set globally for the isolate which unless the worker is run in debug mode is insignificant
+    if (promiseHooks) {
+      // Node >=16.14 only
+      this.stopPromiseHook = promiseHooks.createHook({
```

Comment:
I'd really like to measure the memory/performance penalty we get from this. Seems like it would be possible to delay some of the string processing until we actually needs it, if ever.

Reply:
Sure, I'll leave that up to you.
In any case I feel like it's not super critical ATM.